ATO-732: Pass Reauthentication Failures to RP

BenjaminWCO opened the pull request:
## What

- Refactored Logout Service / Logout Handler to destroy session in Logout Service rather than Logout Handler.
- Pass Reauthentication Failures To RP
- Add "logout reason" extension to LOG_OUT_SUCCESS audit event.

## How to review

Code Review

## Checklist
- [x] Impact on orch and auth mutual dependencies has been checked.

CarlyG55 commented on orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/services/LogoutService.java:
```diff
@@ -149,23 +148,25 @@ public APIGatewayProxyResponseEvent handleLogout(
+            String clientId,
+            URI errorRedirectUri) {
+        var auditUser = createAuditUser(input, session);
+        destroySessions(session);
```
is there a reason we don't check for the existence of the sessions here?

CarlyG55 replied:
Ah if we're already doing a previous session exists check then I'm not too worried

CarlyG55 reviewed (commented):
Nothing major here, just a few little bits. The splitting out of the code into separate commits here was super helpful, made it a lot easier to review - thank you!

BenjaminWCO commented on orchestration-shared/src/test/java/uk/gov/di/orchestration/shared/services/LogoutServiceTest.java:
```diff
@@ -253,20 +249,21 @@ void successfullyReturnsErrorLogoutResponse() {
                         auditUser,
                         AuditService.MetadataPair.pair("rpPairwiseId", rpPairwiseId.get()));
-        verifyNoInteractions(cloudwatchMetricsService);
+        verify(cloudwatchMetricsService).incrementLogout(Optional.of(CLIENT_ID));
```
Ah, so previously destroying the session and incrementing the logout metric was handled in the logout handler class in the case of a logout request but in the relevant handle method in the logout service in the case of an intervention. Given I was adding a third type of logout handle method to the logout service I thought it was best to refactor the service so all three followed the same pattern and did this logic in their respective logout service handle methods. This `verifyNoIntercations` check was presumably here to make sure the metric didn't get incremented twice (once in the logout handler and once in the logout service). It doesn't get called in the logout handler anymore and it actually should get called in the logout service now.

CarlyG55 replied:
I see, makes sense

BenjaminWCO commented on orchestration-shared/src/test/java/uk/gov/di/orchestration/shared/services/LogoutServiceTest.java:
```diff
@@ -328,46 +325,24 @@ void shouldDeleteSessionFromRedisWhenNoCookieExists() {
                         AuditService.MetadataPair.pair("rpPairwiseId", rpPairwiseId.get()));
     }
 
-    @Test
-    void sessionsAreDeletedWhenDestroySessionsIsCalled() {
-
-        setupAdditionalClientSessions();
-
-        logoutService.destroySessions(session);
-
-        verify(backChannelLogoutService)
-                .sendLogoutMessage(
-                        argThat(withClientId("client-id-1")), eq(EMAIL), eq(INTERNAL_SECTOR_URI));
-        verify(backChannelLogoutService)
-                .sendLogoutMessage(
-                        argThat(withClientId("client-id-2")), eq(EMAIL), eq(INTERNAL_SECTOR_URI));
-        verify(backChannelLogoutService)
-                .sendLogoutMessage(
-                        argThat(withClientId("client-id-3")), eq(EMAIL), eq(INTERNAL_SECTOR_URI));
-
-        verify(clientSessionService).deleteStoredClientSession("client-session-id-1");
-        verify(clientSessionService).deleteStoredClientSession("client-session-id-2");
-        verify(clientSessionService).deleteStoredClientSession("client-session-id-3");
-        verify(sessionService, times(1)).deleteSessionFromRedis(SESSION_ID);
-    }
-
```
related to the above, `destroySessions` was public because it was called in the case of a logout handler in the case of a logout request but this logic has been moved into the handle logout method. I should have made destroySessions private. Will fix.

BenjaminWCO replied:
Done

CarlyG55 replied:
I still don't think we should be deleting this test - this is an important bit of logic that we want to make sure works. We should still be doing this testing somewhere

BenjaminWCO replied:
I think most of this logic is covered in other tests but you're right, it does look like none of the other tests check that `backChannelLogoutService.sendLogoutMessage` is called or test for multiple sessions so I think it makes sure these are still covered

BenjaminWCO replied:
Okay, I've added verifications for calls to `backChannelLogoutService.sendLogoutMessage` in the various other tests and a `sessionsAreAllDeletedOnLogout` tests which does much the same as the above but via a `handleLogout` call rather than a call to `logoutService.destroySessions`.

BenjaminWCO commented on oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthenticationCallbackHandler.java:
```diff
@@ -542,26 +539,56 @@ private Map<String, String> buildDimensions(
         return dimensions;
     }
 
+    private Optional<APIGatewayProxyResponseEvent> validateRequest(
```
To be honest I only moved this to a method because sonar didn't like the introduction of a nested try-catch (despite this already being present). A bit of a mouthful but I've renamed to `generateAuthenticationErrorResponseIfRequestInvalid`

CarlyG55 replied:
Makes sense - thanks for that

BenjaminWCO commented on orchestration-shared/src/test/java/uk/gov/di/orchestration/shared/services/LogoutServiceTest.java:
```diff
@@ -328,15 +375,20 @@ void shouldDeleteSessionFromRedisWhenNoCookieExists() {
                 Optional.empty(),
                 rpPairwiseId);
 
-        verify(sessionService, times(0)).deleteSessionFromRedis(SESSION_ID);
```
Although due to the refactor this test is now testing for a session being deleted as it's name suggests, the fact that it wasn't before makes me think this was for some old logic that has now changed. The auditUserWhenNoCookie doesn't seem to trigger any specific logic?

BenjaminWCO replied:
After looking into this it appears this test always asserted deleteSessionFromRedis was not called despite it's name. I'm removing as it's confusingly named and for the current version f the logout service doesn't test any unique logic.

isaac-GDS replied:
Hesitant to delete the test even though its confusingly not properly tested. Since this has gone through a refactor is it that the test hasn't been setup correctly. Seems like the test was purposefully named and should be kept and updated unless someone can confirm this logic change is fine.

isaac-GDS replied:
Should this test be moved into the logoutHandler as thats where the cookie logic is - testing that the `logoutService.destroySession` is still called when no cookie exists.

BenjaminWCO replied:
@billrdunn do you have any thoughts? It looks like this was here since you created this class: https://github.com/govuk-one-login/authentication-api/pull/3647/commits/f374da8b0bf5193ae982c005e38ba94a69323adc

ethanmills replied:
Looks like there is a change in behaviour around emitting metrics though?

BenjaminWCO replied:
As discussed yesterday, will maintain the current behaviour for now, which dosen't delete the session if the cookie is missing. I've opened a new ticket to investigate the desired behaviour.

ethanmills commented on orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/services/LogoutService.java:
```diff
@@ -80,21 +82,15 @@ public LogoutService(
             TxmaAuditUser auditUser,
             Optional<String> clientId,
             Optional<String> rpPairwiseId) {
         LOG.info("Generating logout response using URI: {}", logoutUri);
```
Not sure we should be logging the state here as it's arbitrary RP controlled input and we aren't sure it's not PII

BenjaminWCO replied:
Sure, only logging the path now.

ethanmills commented on orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/services/LogoutService.java:
```diff
@@ -177,21 +192,41 @@ public APIGatewayProxyResponseEvent handleAccountInterventionLogout(
+                LOG_OUT_SUCCESS,
+                auditClientId,
+                auditUser,
+                metadata.toArray(new AuditService.MetadataPair[0]));
```
What's the difference between this and 
```suggestion
                metadata.toArray(AuditService.MetadataPair[]::new));
```
?

BenjaminWCO replied:
Agreed, the second one is better

ethanmills commented:
I might be wrong, but it looks like we were only previously incrementing the cloudwatch metric if there was a session, but now we always do?

BenjaminWCO commented:
> I might be wrong, but it looks like we were only previously incrementing the cloudwatch metric if there was a session, but now we always do?

That's true. I've fixed to remain consistent and added a test.

ethanmills commented on orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/services/LogoutService.java:
```diff
@@ -80,21 +83,17 @@ public LogoutService(
-                        .orElse(logoutUri);
-
-        sendAuditEvent(auditUser, clientId, rpPairwiseId);
+        LOG.info("Generating logout response using URI: {}", logoutUri.getPath());
```
I think this needs host and path to be meaningful?

BenjaminWCO replied:
Yes, fixed
